File: demos/demo_Messi_Streaming.py
"""Build MESSI once and append streaming batches without rebuilding."""

import numpy as np

from daisy import DistanceType, Messi


rng = np.random.default_rng(7)
stream = rng.normal(size=(6000, 96)).astype(np.float32)
queries = rng.normal(size=(5, 96)).astype(np.float32)

index = Messi(DistanceType.L2_SQUARED)
index.setIndexWorkers(2)
index.setSearchWorkers(4)
index.buildIndex(stream[:5000])


def query_and_report(stage, size):
    """The index stays queryable between updates: search after every addition."""
    indices, distances = index.searchIndex(queries, 5)
    print(f"{stage}: MESSI size {size}")
    print("  Query 0 IDs:", indices[0])
    print("  Query 0 distances:", distances[0])


query_and_report("after build", 5000)

index.insert(stream[5000])
query_and_report("after insert", 5001)

index.insertBatch(stream[5001:])
query_and_report("after batch", 6000)

File: demos/demo_Messi_Streaming.cpp
// MESSI streaming: build once, then update the live iSAX tree without rebuilding.

#include "../commons/dataloaders.hpp"
#include "../lib/daisy.hpp"

#include <cstdio>

int main()
{
    const daisy::idx_t dim = 96;
    const daisy::idx_t initial = 5000;
    const daisy::idx_t batch = 1000;
    const daisy::idx_t n_query = 5;
    const daisy::idx_t k = 5;

    float *stream = loadRandomData(initial + batch, dim, 100, true);
    float *query = loadRandomData(n_query, dim, 50, true);

    daisy::Messi search(daisy::DistanceType::L2_SQUARED);
    search.setIndexWorkers(2);
    search.setSearchWorkers(4);
    search.buildIndex(stream, initial, dim);

    daisy::idx_t *indices = new daisy::idx_t[n_query * k];
    float *distances = new float[n_query * k];

    // The index stays queryable between updates: search after every addition.
    auto queryAndReport = [&](const char *stage)
    {
        search.searchIndex(query, n_query, k, indices, distances);
        std::printf("%-16s MESSI contains %llu series. Query 0 kNN: ",
                    stage, search.getNDatabase());
        for (daisy::idx_t j = 0; j < k; ++j)
            std::printf("%llu(%.3f) ", indices[j], distances[j]);
        std::printf("\n");
    };

    queryAndReport("after build");

    search.insert(stream + initial * dim);
    queryAndReport("after insert");

    search.insertBatch(stream + (initial + 1) * dim, batch - 1);
    queryAndReport("after batch");

    delete[] stream;
    delete[] query;
    delete[] indices;
    delete[] distances;
    return 0;
}
